qiskit_optimization/applications/sk_model.py

```diff
@@ -0,0 +1,110 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""An application class for the Sherrington Kirkpatrick (SK) model."""
+
+
+from typing import Union, List
+
+import networkx as nx
+import numpy as np
+from docplex.mp.model import Model
+
+from .graph_optimization_application import OptimizationApplication
+from ..algorithms import OptimizationResult
+from ..problems.quadratic_program import QuadraticProgram
+from ..translators import from_docplex_mp
+
+
+class SKModel(OptimizationApplication):
+    r"""Optimization application of the "Sherrington Kirkpatrick (SK) model" [1].
+
+    The SK Hamiltonian over n spins is given as:
+        :math:`H(x)=-1/\sqrt{n} \sum_{i<j} w_{i,j}x_ix_j`, :math:`where x_i\in\{\pm 1\}`
+    and :math:`w_{i,j}\in\{\pm 1\}` are chosen independently and uniformly at random.
+    Notice that there are other variants e.g., with :math:`w_{i,j}` chosen from the normal distribution
+    with mean 0 and variance 1.
+
+    References:
+        [1]: Dmitry Panchenko. "The Sherrington-Kirkpatrick model: an overview",
+        https://arxiv.org/abs/1211.1094
+    """
+
+    def __init__(self, num_of_sites: int, rng: np.random.RandomState = np.random.RandomState(0)):
```

Comment:
Could you use `Random Generator` instead of legacy `RandomState`? https://numpy.org/doc/stable/reference/random/generator.html

Reply:
I guess this is a similar conversation to above here https://github.com/Qiskit/qiskit-optimization/pull/208#discussion_r675855471

My question first was did this need to be an independent rng. It could internally just use utils\algorithm_globals rng from Terra, as is done in other places in application so this can be seeded from the one central place for over all reproducibility.

if it does need to be independent the next question is do we really need to pass different generators in. Yes they have different properties around the generation but is that needed/wanted? Maybe setting some Optional int seed and having it use default_rng, like algorithm_globals may suffice. If we want generator then maybe a Optional Union of Generator & int whereby a generator can be provided or more simply a seed for default_rng (i.e. do a blend). I cannot help but feel that it getting more complex though and does it really warrant going that far with rng options?

Reply:
Hi, yes, we need to have it reproducible. So in any case it should return deterministic pseudo-random numbers.

Reply:
Hi, yes, we need the rng to be completely independent and produce deterministic pseudo-random numbers.
Yes it is a similar comment as [https://github.com/Qiskit/qiskit-optimization/pull/208#discussion_r675855471](https://github.com/Qiskit/qiskit-optimization/pull/208#discussion_r675855471).

What do you suggest? Objective is we want deterministic pseudo-random numbers independent on anything else.

Reply:
rng is essentially used to set values of edges and the graph is always complete graph. So I would recommend to receive edge weight as `np.ndarray` with shape = (num_of_sites, num_of_sites). Then, users can generate any edge weight as they want. The current design allows only random edge weights.

Related to the recommendation above, I also recommend to use np.ndarray as the internal data structure because the graph is complete.

Reply:
I think that providing edge weights as an array should be a side option only, not the only one.
Ultimately the model/problem is about the random distortions.

Reply:
Regarding not having it as a graph: It is a possibility.
However, having it as a graph have some advantages:
* For some test examples it might be good to be able to draw it with displayed edge weights.
* It is related to MaxCut on random graphs. So by design it is a graph problem. (BTW some variant of SK can have also edge weights 0)

It is subtle, I am not sure where to draw the line. So I picked this as an easy solution.

BTW: recirq has it also represented as a graph.

I would personally stick with the graph representation.

Reply:
If this is a graph problem, could you inherit `GraphOptimizationProblem` instead of `OptimizationProblem`?
Note that `random_graph` is deprecated by #203. But, you can make a utility function as follows, e.g.,
```python
class SKModel(GraphOptimizationProblem):
    def __init__(self, ...):
        ...

    @staticmethod
    def random_instance(num_nodes, rng: Generator) -> SKModel:
        edge_weights = rng.uniform(...)
        return SKModel(edge_weights)
```

This is also a reply to this comment. https://github.com/Qiskit/qiskit-optimization/pull/208#discussion_r677332636

Reply:
It is not a general graph problem. That is the reason I don't inherit from that class. I don't want users to be able to set the graph and so..

Yes that was my request to remove that function. I considered that before.

I think it should contain a graph but not inherit from the GraphOptimizationProblem if it is what it is right now.

Reply:
Could you explain how it is not a general graph problem? Then, we can come up with ideas.

Reply:
I dont want users to input graph or weights. The model should stay as it is.